README
dhcprelya is a lightweight DHCP relay agent written for Yandex company for
high loaded routers with tens interfaces and thousands hosts behind them.

It's fast, independed from any side libraries, and it does not turn on
a promisc mode for listening interfaces. Plugins supported. With plugins
you can implement any functionality. A command line options is compatible
with ISC dhcrelay but extended.

Some ideas was taken from Edwin's dhcprelay (net/dhcprelay in ports collection.
which unfortunately has some shortages) and dhcpdump (net/dhcpdump).

It's distributed under BSD license.

HOW TO RUN
There are two syntax. The first one is a ISC dhcrelay compatible:

dhcprelya [-d] [-p<pidfile>] -A <packet_size> -c <hops_count>
	-i<if1>... -i<ifN> <server1>... <serverN>

-i	- an interfaces list that should be listened.
server	- a servers list where request will be forward to.
-A	- set maximum packet size. Default: 1400. It's more then enough.
-c	- When hops counter in a DHCP packet more then this value,
	  the packet is dropped. It prevets loops. Defaul: 10.

A second one allows to use a configuration file:

dhcprelya [-d] [-p<pidfile>] -f <config.file>

Common options are:
-d	- do not demonize and output debugging info onto terminal.
-p	- use not default pid file. You can run more than one copy of
	  dhcprelya with different pid files.

See dhcprelya.conf-example for format.

If an interface does not exist on this host, it'll be ignored. It's useful
for only one config file for all routers with logical interfaces (vlanNN etc.).

Any questions, bug reports and feature requests are welcome.

Sergey Matveychuk <sem@FreeBSD.org>


HISTORY
2015-07-24:
	4.9 version
	* Fix segfault by using the proper syntax (bapt@)

2014-03-20:
	4.8 version
	* fix max_hops condition

2013-06-19:
	4.7 version
	* fix remote_id parsing in option82_plugin (contributed by svv@sgaice.ru)

2013-04-15:
	4.6 version
	* fix rps_limit converting it to per-interface basis (and removing mutex)

2012-12-03:
	4.5 version
	* destroy() in plugins for smooth shutdown.
	* Change plugin functions declaration. Use more appropriate types.
	* radius_plugin: get rid of freeradius-client library.
	  Use a system libradius instead.
	* radius_plugin: rename bind_interface option to bind_to and allow
	  to specify an IP address there.

2012-11-01:
	4.4 version
	* radius_plugin: send accounting only for DHCPACK packet.

2012-09-02:
	4.3 version
	* Fix options string parsing
	* Minor fix in radius_plugin
	* Fix byte order for eh->ether_type

2012-03-07:
	4.2 version
	* If there is no interface to listen, dhcprelya took 100% CPU.
	* Fix memory leaks.
	* Implement rps_limit opetion to protect a server from a packet storm.
	* Fix checksum calculation in option82_plugin (thanks to Manuel Kasper).
	* Allow specify a DHCP server port. Useful for testing. (Manuel Kasper)

2009-07-16:
	4.1 version
	* In radius plugin: build NAS port as client IP, not random.

2009-06-16:
	4.0 version
	* Be more RFC compliant:
	  + don't change GIADDR if it's not 0.
	  + don't save incoming interface for a client.
	    detect it by GIADDR address. It's simplified
	    code a lot.
	  + don't use INADDR_BROADCAST for answers, but
	    YIADDR instead (excluding bellow).
	  + take into account a BROADCAST flag. If it's on,
	    send an answer as a broadcast (both L2 & L3).
	* Added option82 plugin. RFC 3046 implementation.
	* Add a bind_interface option for radius plugin.

2009-02-13:
	3.0 version
	* Add -A and -c options. They are ISC dhcrelay compatible.
	* Change configuration file format: add sections (ala .ini).
	* Add a plugin architecture.
	* Make log as a plugin. Improve it a lot.

2009-01-13:
	2.3 version
	* Set MAC address in reply packets to MAC address of client.
	  Exotic clients don't get packets with a broadcast MAC.

2009-01-12:
	2.2 version
	* Minor bugfix: buffer size for sendto was a little more.

2008-12-15:
	2.1 version released
	* Close standard descriptors when daemonize.
	* Fix an error in calculation of servers index
	  (when server >2).

2008-09-19:
	2.0 version released
	* It's possible to put your configureation into a config file.
	  The old syntax still can be used for compatibility.

2008-05-26:
	1.3 version released
	* When an interface disabled/removed dhcprelya looped and eated CPU time

2008-04-30:
	1.2 version released
	* Do not turn on promisc mode for interfaces.
	* Do not create PID file when run with -d.
	* A little code clean up.

2008-02-01:
	1.1 version released
	* Optimized. Less system CPU time used.

2007-12-17:
	1.0 version released


dhcp_options.h
/* this list was stolen from The DHCP Handbook by Droms and Lemon, Appendix D */

/* The first comment is the number, the last parameter is if it's verbosed */
const char *dhcp_options[] = {
	 /* 0 */ "pad",
	 /* 1 */ "Subnet mask",	/**/
	 /* 2 */ "Time offset",	/**/
	 /* 3 */ "Routers",	/**/
	 /* 4 */ "Time server",	/**/
	 /* 5 */ "Name server",	/**/
	 /* 6 */ "DNS server",	/**/
	 /* 7 */ "Log server",	/**/
	 /* 8 */ "Cookie server",	/**/
	 /* 9 */ "LPR server",	/**/
	 /* 10 */ "Impress server",	/**/
	 /* 11 */ "Resource location server",	/**/
	 /* 12 */ "Host name",	/**/
	 /* 13 */ "Boot file size",	/**/
	 /* 14 */ "Merit dump file",	/**/
	 /* 15 */ "Domainname",	/**/
	 /* 16 */ "Swap server",/**/
	 /* 17 */ "Root path",	/**/
	 /* 18 */ "Extensions path",	/**/
	 /* 19 */ "IP forwarding",	/**/
	 /* 20 */ "Non-local source routing",	/**/
	 /* 21 */ "Policy filter",	/**/
	 /* 22 */ "Maximum datagram reassembly size",	/**/
	 /* 23 */ "Default IP TTL",	/**/
	 /* 24 */ "Path MTU aging timeout",	/**/
	 /* 25 */ "Path MTU plateau table",	/**/
	 /* 26 */ "Interface MTU",	/**/
	 /* 27 */ "All subnets local",	/**/
	 /* 28 */ "Broadcast address",	/**/
	 /* 29 */ "Perform mask discovery",	/**/
	 /* 30 */ "Mask supplier",	/**/
	 /* 31 */ "Perform router discovery",	/**/
	 /* 32 */ "Router solicitation",	/**/
	 /* 33 */ "Static route",	/**/
	 /* 34 */ "Trailer encapsulation",	/**/
	 /* 35 */ "ARP cache timeout",	/**/
	 /* 36 */ "Ethernet encapsulation",	/**/
	 /* 37 */ "TCP default TTL",	/**/
	 /* 38 */ "TCP keepalive interval",	/**/
	 /* 39 */ "TCP keepalive garbage",	/**/
	 /* 40 */ "NIS domain",	/**/
	 /* 41 */ "NIS servers",/**/
	 /* 42 */ "NTP servers",/**/
	 /* 43 */ "Vendor specific info",	/**/
	 /* 44 */ "NetBIOS name server",	/**/
	 /* 45 */ "NetBIOS datagram distribution server",	/**/
	 /* 46 */ "NetBIOS node type",	/**/
	 /* 47 */ "NetBIOS scope",	/**/
	 /* 48 */ "X Window System font server",	/**/
	 /* 49 */ "X Window System display server",	/**/
	 /* 50 */ "Request IP address",	/**/
	 /* 51 */ "IP address leasetime",	/**/
	 /* 52 */ "Option overload",	/**/
	 /* 53 */ "DHCP message type",	/**/
	 /* 54 */ "Server identifier",	/**/
	 /* 55 */ "Parameter Request List",	/**/
	 /* 56 */ "Message",	/**/
	 /* 57 */ "Maximum DHCP message size",	/**/
	 /* 58 */ "T1",		/**/
	 /* 59 */ "T2",		/**/
	 /* 60 */ "Vendor class identifier",	/**/
	 /* 61 */ "Client-identifier",	/**/
	 /* 62 */ "Netware/IP domain name",	/**/
	 /* 63 */ "Netware/IP domain information",	/**/
	 /* 64 */ "NIS+ domain",/**/
	 /* 65 */ "NIS+ servers",	/**/
	 /* 66 */ "TFTP server name",	/**/
	 /* 67 */ "Bootfile name",	/**/
	 /* 68 */ "Mobile IP home agent",	/**/
	 /* 69 */ "SMTP server",/**/
	 /* 70 */ "POP3 server",/**/
	 /* 71 */ "NNTP server",/**/
	 /* 72 */ "WWW server",	/**/
	 /* 73 */ "Finger server",	/**/
	 /* 74 */ "IRC server",	/**/
	 /* 75 */ "StreetTalk server",	/**/
	 /* 76 */ "StreetTalk directory assistance server",	/**/
	 /* 77 */ "User-class Identification",
	 /* 78 */ "SLP-directory-agent",
	 /* 79 */ "SLP-service-scope",
	 /* 80 */ "Naming Authority",
	 /* 81 */ "Client FQDN",/**/
	 /* 82 */ "Relay Agent Information",
	 /* 83 */ "Agent Remote ID",
	 /* 84 */ "Agent Subnet Mask",
	 /* 85 */ "NDS server",	/**/
	 /* 86 */ "NDS tree name",	/**/
	 /* 87 */ "NDS context",/**/
	 /* 88 */ "IEEE 1003.1 POSIX",
	 /* 89 */ "FQDN",
	 /* 90 */ "Authentication",
	 /* 91 */ "Vines TCP/IP",
	 /* 92 */ "Server Selection",
	 /* 93 */ "Client System",
	 /* 94 */ "Client NDI",
	 /* 95 */ "LDAP",
	 /* 96 */ "IPv6 Transitions",
	 /* 97 */ "UUID/GUID",
	 /* 98 */ "UPA servers",
	 /* 99 */ "???",
	 /* 100 */ "Printer Name",
	 /* 101 */ "MDHCP",
	 /* 102 */ "???",
	 /* 103 */ "???",
	 /* 104 */ "???",
	 /* 105 */ "???",
	 /* 106 */ "???",
	 /* 107 */ "???",
	 /* 108 */ "Swap Path",
	 /* 109 */ "???",
	 /* 110 */ "IPX Compatability",
	 /* 111 */ "???",
	 /* 112 */ "Netinfo Address",
	 /* 113 */ "Netinfo Tag",
	 /* 114 */ "URL",
	 /* 115 */ "DHCP Failover",
	 /* 116 */ "DHCP Autoconfiguration",
	 /* 117 */ "Name Service Search",
	 /* 118 */ "Subnet selection",
	 /* 119 */ "Domain Search",
	 /* 120 */ "SIP Servers DHCP Option",
	 /* 121 */ "Classless Static Route",
	 /* 122 */ "???",
	 /* 123 */ "???",
	 /* 124 */ "???",
	 /* 125 */ "???",
	 /* 126 */ "Extension",
	 /* 127 */ "Extension",
	 /* 128 */ "???",
	 /* 129 */ "???",
	 /* 130 */ "???",
	 /* 131 */ "???",
	 /* 132 */ "???",
	 /* 133 */ "???",
	 /* 134 */ "???",
	 /* 135 */ "???",
	 /* 136 */ "???",
	 /* 137 */ "???",
	 /* 138 */ "???",
	 /* 139 */ "???",
	 /* 140 */ "???",
	 /* 141 */ "???",
	 /* 142 */ "???",
	 /* 143 */ "???",
	 /* 144 */ "HP - TFTP file",
	 /* 145 */ "???",
	 /* 146 */ "???",
	 /* 147 */ "???",
	 /* 148 */ "???",
	 /* 149 */ "???",
	 /* 150 */ "CiscoCallManagerTFTP",
	 /* 151 */ "???",
	 /* 152 */ "???",
	 /* 153 */ "???",
	 /* 154 */ "???",
	 /* 155 */ "???",
	 /* 156 */ "???",
	 /* 157 */ "???",
	 /* 158 */ "???",
	 /* 159 */ "???",
	 /* 160 */ "???",
	 /* 161 */ "???",
	 /* 162 */ "???",
	 /* 163 */ "???",
	 /* 164 */ "???",
	 /* 165 */ "???",
	 /* 166 */ "???",
	 /* 167 */ "???",
	 /* 168 */ "???",
	 /* 169 */ "???",
	 /* 170 */ "???",
	 /* 171 */ "???",
	 /* 172 */ "???",
	 /* 173 */ "???",
	 /* 174 */ "???",
	 /* 175 */ "???",
	 /* 176 */ "???",
	 /* 177 */ "???",
	 /* 178 */ "???",
	 /* 179 */ "???",
	 /* 180 */ "???",
	 /* 181 */ "???",
	 /* 182 */ "???",
	 /* 183 */ "???",
	 /* 184 */ "???",
	 /* 185 */ "???",
	 /* 186 */ "???",
	 /* 187 */ "???",
	 /* 188 */ "???",
	 /* 189 */ "???",
	 /* 190 */ "???",
	 /* 191 */ "???",
	 /* 192 */ "???",
	 /* 193 */ "???",
	 /* 194 */ "???",
	 /* 195 */ "???",
	 /* 196 */ "???",
	 /* 197 */ "???",
	 /* 198 */ "???",
	 /* 199 */ "???",
	 /* 200 */ "???",
	 /* 201 */ "???",
	 /* 202 */ "???",
	 /* 203 */ "???",
	 /* 204 */ "???",
	 /* 205 */ "???",
	 /* 206 */ "???",
	 /* 207 */ "???",
	 /* 208 */ "???",
	 /* 209 */ "???",
	 /* 210 */ "Authenticate",
	 /* 211 */ "???",
	 /* 212 */ "???",
	 /* 213 */ "???",
	 /* 214 */ "???",
	 /* 215 */ "???",
	 /* 216 */ "???",
	 /* 217 */ "???",
	 /* 218 */ "???",
	 /* 219 */ "???",
	 /* 220 */ "???",
	 /* 221 */ "???",
	 /* 222 */ "???",
	 /* 223 */ "???",
	 /* 224 */ "???",
	 /* 225 */ "???",
	 /* 226 */ "???",
	 /* 227 */ "???",
	 /* 228 */ "???",
	 /* 229 */ "???",
	 /* 230 */ "???",
	 /* 231 */ "???",
	 /* 232 */ "???",
	 /* 233 */ "???",
	 /* 234 */ "???",
	 /* 235 */ "???",
	 /* 236 */ "???",
	 /* 237 */ "???",
	 /* 238 */ "???",
	 /* 239 */ "???",
	 /* 240 */ "???",
	 /* 241 */ "???",
	 /* 242 */ "???",
	 /* 243 */ "???",
	 /* 244 */ "???",
	 /* 245 */ "???",
	 /* 246 */ "???",
	 /* 247 */ "???",
	 /* 248 */ "???",
	 /* 249 */ "MSFT - Classless route",
	 /* 250 */ "???",
	 /* 251 */ "???",
	 /* 252 */ "MSFT - WinSock Proxy Auto Detect",
	 /* 253 */ "???",
	 /* 254 */ "???",
 /* 255 */ "End"};


const char *dhcp_message_types[] = {
	"wrong specified",
	 /* 1 */ "DHCPDISCOVER",
	 /* 2 */ "DHCPOFFER",
	 /* 3 */ "DHCPREQUEST",
	 /* 4 */ "DHCPDECLINE",
	 /* 5 */ "DHCPACK",
	 /* 6 */ "DHCPNAK",
	 /* 7 */ "DHCPRELEASE",
	 /* 8 */ "DHCPINFORM"
};

const char *netbios_node_type[] = {
	 /* 0 */ "none",
	 /* 1 */ "B-node",
	 /* 2 */ "P-node",
	 /* 3 */ "",
	 /* 4 */ "M-node",
	 /* 5 */ "",
	 /* 6 */ "",
	 /* 7 */ "",
	 /* 8 */ "H-node"
};

const char *option_overload[] = {
	 /* 0 * */ "unspecified",
	 /* 1 */ "file field holds options",
	 /* 2 */ "sname field holds options",
	 /* 3 */ "file and sname field holds options"
};

const char *enabledisable[] = {
	 /* 0 */ "disabled",
	 /* 1 */ "enabled"
};

const char *ethernet_encapsulation[] = {
	 /* 0 */ "Ethernet version 2",
	 /* 1 */ "IEEE 802.3"
};

const char *operands[] = {
	 /* 0 */ "wrong specified",
	 /* 1 */ "BOOTPREQUEST",
	 /* 2 */ "BOOTPREPLY"
};

/* From RFC 3046 */
const char *relayagent_suboptions[] = {
	 /* 0 */ "<unknown-0>",
	 /* 1 */ "Circuit-ID",
	 /* 2 */ "Remote-ID"
};

/* Copied from RFC1700 */
const char *htypes[] = {
	 /* 0 */ "wrong specified",
	 /* 1 */ "Ethernet",
	 /* 2 */ "Experimental Ethernet",
	 /* 3 */ "Amateur Radio AX.25",
	 /* 4 */ "Proteon ProNET Token Ring",
	 /* 5 */ "Chaos",
	 /* 6 */ "IEEE 802 Networks",
	 /* 7 */ "ARCNET",
	 /* 8 */ "Hyperchannel",
	 /* 9 */ "Lanstar",
	 /* 10 */ "Autonet Short Address",
	 /* 11 */ "LocalTalk",
	 /* 12 */ "LocalNet",
	 /* 13 */ "Ultra link",
	 /* 14 */ "SMDS",
	 /* 15 */ "Frame Relay",
	 /* 16 */ "ATM",
	 /* 17 */ "HDLC",
	 /* 18 */ "Fibre Channel",
	 /* 19 */ "ATM",
	 /* 20 */ "Serial Line",
	 /* 21 */ "ATM"
};
